executor/prepared_test.go

```diff
@@ -104,6 +109,64 @@ func (s *testSuite1) TestPrepareStmtAfterIsolationReadChange(c *C) {
 	c.Assert(tk.Se.GetSessionVars().PreparedStmts[1].(*plannercore.CachedPrepareStmt).NormalizedPlan, Equals, "")
 }
 
+type mockSessionManager2 struct {
+	se     session.Session
+	killed bool
+}
+
+func (sm *mockSessionManager2) ShowProcessList() map[uint64]*util.ProcessInfo {
+	pl := make(map[uint64]*util.ProcessInfo)
+	if pi, ok := sm.GetProcessInfo(0); ok {
+		pl[pi.ID] = pi
+	}
+	return pl
+}
+
+func (sm *mockSessionManager2) GetProcessInfo(id uint64) (pi *util.ProcessInfo, notNil bool) {
+	pi = sm.se.ShowProcess()
+	if pi != nil {
+		notNil = true
+	}
+	return
+}
+func (sm *mockSessionManager2) Kill(connectionID uint64, query bool) {
+	sm.killed = true
+	atomic.StoreUint32(&sm.se.GetSessionVars().Killed, 1)
+}
+func (sm *mockSessionManager2) UpdateTLSConfig(cfg *tls.Config) {}
+
+var _ = SerialSuites(&testSuite12{&baseTestSuite{}})
+
+type testSuite12 struct {
+	*baseTestSuite
+}
+
+func (s *testSuite12) TestPreparedStmtWithHint(c *C) {
+	// see https://github.com/pingcap/tidb/issues/18535
+	store, dom, err := newStoreWithBootstrap()
+	c.Assert(err, IsNil)
+	defer func() {
+		store.Close()
+		dom.Close()
+	}()
+
+	se, err := session.CreateSession4Test(store)
+	c.Assert(err, IsNil)
+
+	sm := &mockSessionManager2{
+		se: se,
+	}
+	_, err = se.Execute(context.Background(), "prepare stmt from \"select /*+ max_execution_time(100) */ sleep(10)\"")
+	c.Assert(err, IsNil)
+	se.SetSessionManager(sm)
+	go dom.ExpensiveQueryHandle().SetSessionManager(sm).Run()
+	rs, err := se.Execute(context.Background(), "execute stmt")
+	c.Check(err, IsNil)
+	tk := testkit.NewTestKit(c, store)
+	tk.ResultSetToResult(rs[0], Commentf("%v", rs)).Check(testkit.Rows("1"))
+	c.Check(sm.killed, Equals, true)
```

Comment:
Nitpick: 
1. I think we should separate init resources and test.
```
tk := testkit.NewTestKit(c, store)
se.SetSessionManager(sm)
go dom.ExpensiveQueryHandle().SetSessionManager(sm).Run()
.....
se.Execute....
tk.ResultSetToResult....
```
2.  Why not use tk.MustExec, tk.MustQuery?

Reply:
I'll try using tk.MustExec, however for seperat init resources and test, I guess you mean we should put the declaration of the mockSessionManager at another place, but it is only used specifically here as a helper structure, I don't see a better place. Therefore I would like to leave it here, and if it is used in another place in the future, we could then refactoring it.